src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala

```diff
@@ -0,0 +1,351 @@
+      stash()
+      unstashAll()
+      context.become(active)
+    case message: Any => stash()
+  }
+
+  private[this] def active: Receive = LoggingReceive.withLabel("active") {
+    Seq(
+      receiveStop,
+      receiveDelayUpdate,
+      receiveTaskLaunchNotification,
+      receiveTaskStatusUpdate,
+      receiveGetCurrentCount,
+      receiveAddCount,
+      receiveProcessOffers
+    ).reduce(_.orElse[Any, Unit](_))
+  }
+
+  private[this] def receiveWaitingForInFlight: Receive = LoggingReceive.withLabel("waitingForInFlight") {
+    case notification: TaskLaunchNotification =>
+      receiveTaskLaunchNotification(notification)
+      waitingForInFlight()
+
+    case "waitingForInFlight" => sender() ! "waitingForInFlight" // for testing
+  }
+
+  private[this] def receiveStop: Receive = {
+    case AppTaskLauncherActor.Stop => waitingForInFlight()
+  }
+
+  private[this] def waitingForInFlight(): Unit = {
+    if (inFlightTaskLaunches.isEmpty) {
+      context.stop(self)
+    }
+    else {
+      context.become(receiveWaitingForInFlight)
+    }
+  }
+
+  /**
+    * Receive rate limiter updates.
+    */
+  private[this] def receiveDelayUpdate: Receive = {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      if (backOffUntil.forall(_ < delayUntil)) {
+        import context.dispatcher
+        val now: Timestamp = clock.now()
+        recheckBackOff.foreach(_.cancel())
+        recheckBackOff = Some(
+          context.system.scheduler.scheduleOnce(now until delayUntil, self, RecheckIfBackOffUntilReached)
+        )
+      }
+
+      backOffUntil = Some(delayUntil)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+      log.debug("After delay update {}", status)
+
+    case RecheckIfBackOffUntilReached => OfferMatcherRegistration.manageOfferMatcherStatus()
+  }
+
+  private[this] def receiveTaskLaunchNotification: Receive = {
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) if inFlightTaskLaunches(taskInfo.getTaskId) =>
+      // This task is not yet known to mesos, so there will be no event that removes
+      // it automatically from the taskTracker.
+      taskTracker.terminated(app.id, taskInfo.getTaskId.getValue)
+      removeTask(taskInfo.getTaskId)
+      tasksToLaunch += 1
+      log.info(
+        "Task launch for '{}' was denied, reason '{}', rescheduling. {}",
+        taskInfo.getTaskId.getValue, reason, status)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) => // ignore
+
+    case TaskLaunchSourceDelegate.TaskLaunchAccepted(taskInfo) =>
+      inFlightTaskLaunches -= taskInfo.getTaskId
+      log.info("Task launch for '{}' was accepted. {}", taskInfo.getTaskId.getValue, status)
+  }
+
+  private[this] def receiveTaskStatusUpdate: Receive = {
+    case TaskStatusUpdate(_, taskId, MarathonTaskStatus.Terminal(_)) =>
+      log.debug("task '{}' finished", taskId.getValue)
```

Comment:
Incorrect. We have to update our own list.

Our own list should be the same as the task tracker list, except:

* we do not update status because we don't need it yet
* we also include in-flight tasks which are not necessarily yet in the task tracker